### PR TITLE
Replay from dead/forked node: Introduce `NodeStateDump`

### PR DESCRIPTION
**Pull request by oxade**
## Description 

This introduces `NodeStateDump` which is the set of objects and other metadata required to reproduce a fork in replay.
The created state is a few MB, so not too concerning. Json was so the data is human readable.

Most of the logic in this PR is in getting the replay system to load from state dump.

Sample config
```
state-debug-dump-config:
  dump-file-directory: /tmp/
```

## Test Plan 

Tested by simulating forks and replaying from dumps.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-config/src/builder.rs [hunk 1]
```diff
@@ -36,7 +37,6 @@ use sui_types::crypto::{
 };
 use sui_types::multiaddr::Multiaddr;
 use sui_types::object::Object;
-
```
Comment (mwtian) [inline crates/sui-config/src/builder.rs:39]: nit: unnecessary to remove this

File: crates/sui-types/src/temporary_store.rs [hunk 2]
```diff
@@ -1494,7 +1501,14 @@ impl<S: ChildObjectResolver> Storage for TemporaryStore<S> {
 
 impl<S: BackingPackageStore> BackingPackageStore for TemporaryStore<S> {
     fn get_package_object(&self, package_id: &ObjectID) -> SuiResult<Option<Object>> {
-        self.store.get_package_object(package_id)
+        self.store.get_package_object(package_id).map(|obj| {
+            // Track object but leave unchanged
+            if let Some(v) = obj.clone() {
+                // Can this lock ever block execution?
+                self.runtime_read_objects.write().insert(*package_id, v);
```
Comment (mwtian) [inline crates/sui-types/src/temporary_store.rs:1508]: Are `runtime_read_objects` inserted elsewhere?

File: crates/sui-config/src/builder.rs (hunk 1 above)
Comment (oxade) [reply crates/sui-config/src/builder.rs:39]: Yeah not intentional; my formatter has been acting weird lately. Will revert.

File: crates/sui-types/src/temporary_store.rs (hunk 2 above)
Comment (oxade) [reply crates/sui-types/src/temporary_store.rs:1508]: Not always, especially for Move calls.

File: crates/sui-config/tests/snapshots/snapshot_tests__network_config_snapshot_matches.snap [hunk 3]
```diff
@@ -90,6 +90,8 @@ validator_configs:
       shared_object_disabled: false
       user_transaction_disabled: false
     certificate-deny-config: {}
+    state-debug-dump-config:
+      dump-file-directory: /var/folders/3m/hnrdf4892mj0df4tqf55pm_m0000gn/T/
```
Comment (mystenmark) [inline crates/sui-config/tests/snapshots/snapshot_tests__network_config_snapshot_matches.snap:94]: this of course won't pass tests - i like @mwtian's suggestion of defaulting it to empty and then returning a tempdir if it is unset.

Comment (oxade) [reply crates/sui-config/tests/snapshots/snapshot_tests__network_config_snapshot_matches.snap:94]: Yup, fixing that right now.

File: crates/sui-config/src/node.rs [hunk 4]
```diff
@@ -736,6 +739,21 @@ impl AuthorityKeyPairWithPath {
...
+#[serde(rename_all = "kebab-case")]
+pub struct StateDebugDumpConfig {
+    #[serde(default)]
+    pub dump_file_directory: PathBuf,
+}
+
+impl Default for StateDebugDumpConfig {
+    fn default() -> Self {
+        Self {
+            dump_file_directory: std::env::temp_dir(),
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-config/src/node.rs:752]: Good idea. Fixing

File: crates/sui-core/src/authority.rs [hunk 5]
```diff
@@ -925,6 +1007,38 @@ impl AuthorityState {
 
         if let Some(expected_effects_digest) = expected_effects_digest {
             if effects.digest() != expected_effects_digest {
+                // We dont want to mask the error here, so we log it and continue.
+                let dump_dir = self
+                    .debug_dump_config
+                    .dump_file_directory
+                    .as_ref()
+                    .cloned()
+                    .unwrap_or(std::env::temp_dir());
```
Comment (lxfind) [inline crates/sui-core/src/authority.rs:1016]: These can probably be moved to inside state_dump

File: crates/sui-config/src/node.rs [hunk 6]
```diff
@@ -736,6 +739,13 @@ impl AuthorityKeyPairWithPath {
     }
 }
 
+#[derive(Clone, Debug, Deserialize, Serialize, Default)]
+#[serde(rename_all = "kebab-case")]
+pub struct StateDebugDumpConfig {
```
Comment (lxfind) [inline crates/sui-config/src/node.rs:744]: Some doc comment on the purpose of this config would be good

File: crates/sui-core/src/authority.rs [hunk 7]
```diff
@@ -866,6 +870,84 @@ impl AuthorityState {
             .check_owned_object_locks_exist(owned_object_refs)
     }
 
+    /// Make sure not to leak any private info here
```
Comment (lxfind) [inline crates/sui-core/src/authority.rs:873]: Doc comment this function, instead of just the note

File: crates/sui-core/src/authority.rs [hunk 8]
```diff
@@ -866,6 +870,84 @@ impl AuthorityState {
...
+    pub(crate) async fn state_dump(
```
Comment (lxfind) [inline crates/sui-core/src/authority.rs:874]: Might be better called something like debug_dump_transaction_state

File: crates/sui-core/src/authority.rs [hunk 9]
```diff
@@ -866,6 +870,84 @@ impl AuthorityState {
...
+        }
+
+        // Objects and packages read at runtime
+        let mut runtime_reads = Vec::new();
+        for obj in inner_temporary_store.runtime_read_objects.values() {
+            runtime_reads.push(obj.clone());
+        }
+
+        // All other input objects should already be in `inner_temporary_store.objects`
+        let file_path = NodeStateDump {
```
Comment (lxfind) [inline crates/sui-core/src/authority.rs:930]: Might be cleaner to define a constructor in NodeStateDump that takes 2 stores + tx/effects which would propagate all fields inside

File: crates/sui-core/src/authority.rs [hunk 10]
```diff
@@ -866,6 +870,84 @@ impl AuthorityState {
...
+            if let Some(w) = self.get_object(&sys_package_id).await? {
+                relevant_system_packages.push(w)
+            }
+        }
+
+        // Record all the shared objects
+        let mut shared_objects = Vec::new();
+        for (id, ver, _) in effects.shared_objects() {
+            if let Some(w) = self.database.get_object_by_key(id, *ver)? {
+                shared_objects.push(w)
```
Comment (lxfind) [inline crates/sui-core/src/authority.rs:903]: Are these not going to be covered in runtime_read_objects?
Similar question for loaded_child_objects

**Review by lxfind — COMMENTED** [review]
Some nits + questions, overall LGTM

**Review by lxfind — COMMENTED** [review]
Some nits + questions, overall LGTM

**Review by lxfind — COMMENTED** [review]
Some nits + questions, overall LGTM

**Review by lxfind — COMMENTED** [review]
Some nits + questions, overall LGTM

**Review by lxfind — COMMENTED** [review]
Some nits + questions, overall LGTM

**Review by lxfind — COMMENTED** [review]
Some nits + questions, overall LGTM

**Review by lxfind — COMMENTED** [review]
Some nits + questions, overall LGTM

**Comment by lxfind** [comment]
lol I do no know why github created 5 copies of my comment

File: crates/sui-core/src/authority.rs (hunk 10 above)
Comment (oxade) [reply crates/sui-core/src/authority.rs]: `runtime_read_objects`
Not always/ In some cases dependent packages are fetched even thought they're not in the input objects
`loaded_child_objects `
When a child object is just read but not mutated, the object does not show up anywhere else. Had to add this logic explicitly for debugging weeks back.

**Comment by oxade** [comment]
> lol I do no know why github created 10 copies of my comment

Strange. It's all good.
Thanks for review. All comments are addressed now.

**Review by lxfind — APPROVED** [review]

File: crates/sui-replay/src/data_fetcher.rs [hunk 11]
```diff
@@ -189,6 +329,71 @@ impl DataFetcher for RemoteFetcher {
...
+
+    /// Very testnet specific
+    /// This function is testnet specific and will be extended for mainnet later
+    async fn get_epoch_start_timestamp_and_rgp(
+        &self,
+        epoch_id: u64,
+        is_testnet: bool,
+    ) -> Result<(u64, u64), LocalExecError> {
+        // Hack for testnet: for epoch in range [3, 742), we have no data, but no user TX was executed, so return dummy
+        if (is_testnet) && (2 < epoch_id) && (epoch_id < 742) {
```
Comment (tnowacki) [inline crates/sui-replay/src/data_fetcher.rs:341]:
```suggestion
        if is_testnet && (2 < epoch_id) && (epoch_id < 742) {
```

**Pull request merged**